crypto/src/fixpoint.rs

```diff
@@ -0,0 +1,199 @@
+use std::{
+    fmt::Display,
+    ops::{Add, Div, Mul, Sub},
+};
+
+mod div;
+
+use ethnum::U256;
+
+use crate::Amount;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct U128x128(U256);
+
+impl From<u128> for U128x128 {
+    fn from(value: u128) -> Self {
+        Self(U256([0, value]))
+    }
+}
+
+impl From<Amount> for U128x128 {
+    fn from(value: Amount) -> Self {
+        let value = value.value() as u128;
+        value.into()
+    }
+}
+
+impl U128x128 {
+    pub fn is_integral(&self) -> bool {
+        let fractional_word = self.0.into_words().1;
+        fractional_word == 0
+    }
+
+    pub fn rounding(self) -> Option<Self> {
+        let one_half: u128 = 1 << 127;
+        let (integral, fractional) = self.0.into_words();
+        if fractional >= one_half {
+            let Some(integral) = integral.checked_add(1) else {
+                return None;
+            };
+            Some(Self(U256::from_words(integral, 0u128)))
+        } else {
+            Some(self.truncate())
+        }
+    }
```

Comment:
I think this would be better as two functions:
```
pub fn round_up(&self) -> Self { ... }
pub fn round_down(&self) -> Self { ... }
```
so that the caller is always choosing whether to round up or down, since that will probably depend on the use case. For instance, we'll generally want to round one direction or the other when working with the DEX, so that we're always burning rounding error rather than minting dust, and which direction is the right one might depend on the call location.

Reply:
Fixed in https://github.com/penumbra-zone/penumbra/pull/2008/commits/a10d983a26e473fd8e003aa01cdfbae4fac466f8